### PR TITLE
Added an warn alarm for when the daily rate limit has been reached

### PR DESCRIPTION
**Pull request by jimleroyer**
## Changes

Added a warning alarm for when the daily rate limit has been reached.

## Issue

https://trello.com/c/62gscLJH/169-graph-the-services-going-over-the-daily-rate-limit

## Test 

* A similar alarm has been created in staging and has been tested. This is the infrastructure code to automate the alarm creation.
* Once this has been merged and deployed to staging, I will trigger the alarm manually with a test service.

### CODE REVIEW

File: aws/common/cloudwatch_alarms.tf [hunk 1]
```diff
@@ -192,3 +192,17 @@ resource "aws_cloudwatch_metric_alarm" "no-emails-sent-5-minutes-critical" {
   treat_missing_data  = "breaching"
   alarm_actions       = [aws_sns_topic.notification-canada-ca-alert-critical.arn]
 }
+
+resource "aws_cloudwatch_metric_alarm" "daily-rate-limit-error-1-minute-warning" {
+  alarm_name          = "daily-rate-limit-error-1-minute-warning"
+  alarm_description   = "The daily rate limit has been reached."
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_alarms.tf:198]: We've got a ton of various rate limits on Notify. Could you update slug/name/description to reflect that this is a daily rate limit for _services_?

File: aws/common/cloudwatch_alarms.tf [hunk 2]
```diff
@@ -192,3 +192,17 @@ resource "aws_cloudwatch_metric_alarm" "no-emails-sent-5-minutes-critical" {
   treat_missing_data  = "breaching"
   alarm_actions       = [aws_sns_topic.notification-canada-ca-alert-critical.arn]
 }
+
+resource "aws_cloudwatch_metric_alarm" "daily-rate-limit-error-1-minute-warning" {
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_alarms.tf:196]: Could you move this alarm to the other file? https://github.com/cds-snc/notification-terraform/blob/main/aws/eks/cloudwatch_log.tf Other log metric alarms are there

File: aws/common/cloudwatch_alarms.tf [hunk 3]
```diff
@@ -192,3 +192,17 @@ resource "aws_cloudwatch_metric_alarm" "no-emails-sent-5-minutes-critical" {
...
+  alarm_name          = "daily-rate-limit-error-1-minute-warning"
+  alarm_description   = "The daily rate limit has been reached."
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = "over-daily-rate-limit"
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_alarms.tf:201]: You should be able to use `aws_cloudwatch_log_metric_filter.over-daily-rate-limit.name` after moving to the other file

File: aws/common/cloudwatch_alarms.tf [hunk 4]
```diff
@@ -192,3 +192,17 @@ resource "aws_cloudwatch_metric_alarm" "no-emails-sent-5-minutes-critical" {
...
+  namespace           = "LogMetrics"
+  period              = "60"
+  statistic           = "Sum"
+  threshold           = 1
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_alarms.tf:205]: This is not a very important alarm for now, what do you think about changing to period to 300s to evaluate this alarm less frequently?

File: aws/common/cloudwatch_alarms.tf [hunk 5]
```diff
@@ -192,3 +192,17 @@ resource "aws_cloudwatch_metric_alarm" "no-emails-sent-5-minutes-critical" {
...
+  treat_missing_data  = "notBreaching"
```
Comment (AntoineAugusti) [inline aws/common/cloudwatch_alarms.tf:206]: 👍

**Review by AntoineAugusti — CHANGES_REQUESTED** [review]

File: aws/common/cloudwatch_alarms.tf (hunk 2 above)
Comment (AntoineAugusti) [reply aws/common/cloudwatch_alarms.tf:196]: Sorry, wrong link https://github.com/cds-snc/notification-terraform/blob/main/aws/eks/cloudwatch_alarms.tf

File: aws/common/cloudwatch_alarms.tf (hunk 1 above)
Comment (jimleroyer) [reply aws/common/cloudwatch_alarms.tf:198]: Shall we update the filter `over-daily-rate-limit` in the `LogMetrics` namespace too then?

File: aws/common/cloudwatch_alarms.tf (hunk 4 above)
Comment (jimleroyer) [reply aws/common/cloudwatch_alarms.tf:205]: Good with me.

File: aws/eks/cloudwatch_alarms.tf [hunk 6]
```diff
@@ -290,3 +290,17 @@ resource "aws_cloudwatch_metric_alarm" "ddos-detected-load-balancer-critical" {
     ResourceArn = aws_shield_protection.notification-canada-ca.resource_arn
   }
 }
+
+resource "aws_cloudwatch_metric_alarm" "daily-service-rate-limit-error-1-minute-warning" {
+  alarm_name          = "daily-service-rate-limit-error-1-minute-warning"
+  alarm_description   = "The daily rate limit has been reached."
```
Comment (AntoineAugusti) [inline aws/eks/cloudwatch_alarms.tf:296]:
```suggestion
  alarm_description   = "The daily rate limit has been reached for a service"
```

File: aws/eks/cloudwatch_alarms.tf [hunk 7]
```diff
@@ -290,3 +290,17 @@ resource "aws_cloudwatch_metric_alarm" "ddos-detected-load-balancer-critical" {
     ResourceArn = aws_shield_protection.notification-canada-ca.resource_arn
   }
 }
+
+resource "aws_cloudwatch_metric_alarm" "daily-service-rate-limit-error-1-minute-warning" {
+  alarm_name          = "daily-service-rate-limit-error-1-minute-warning"
```
Comment (AntoineAugusti) [inline aws/eks/cloudwatch_alarms.tf:295]:
```suggestion
resource "aws_cloudwatch_metric_alarm" "daily-service-rate-limit-error-5-minutes-warning" {
  alarm_name          = "daily-service-rate-limit-error-5-minutes-warning"
```

File: aws/eks/cloudwatch_alarms.tf [hunk 8]
```diff
@@ -290,3 +290,17 @@ resource "aws_cloudwatch_metric_alarm" "ddos-detected-load-balancer-critical" {
...
+  alarm_description   = "The daily rate limit has been reached."
+  comparison_operator = "GreaterThanOrEqualToThreshold"
+  evaluation_periods  = "1"
+  metric_name         = aws_cloudwatch_log_metric_filter.over-daily-rate-limit.name
+  namespace           = "LogMetrics"
```
Comment (AntoineAugusti) [inline aws/eks/cloudwatch_alarms.tf:300]: You should be able to use the `namespace` attribute of `aws_cloudwatch_log_metric_filter.over-daily-rate-limit`

**Review by AntoineAugusti — APPROVED** [review]
Thanks for setting up this alarm @jimleroyer! 👏 

Made a few minor suggestions, feel free to merge after incorporating these. I would be in favour of revisiting the log metric name as well to put a blatant _service_ keyword in there, if you could do that.

Please update the alarm sheet one last time just before merging 👍

File: aws/eks/cloudwatch_alarms.tf (hunk 8 above)
Comment (jimleroyer) [reply aws/eks/cloudwatch_alarms.tf]: Hmm it makes sense but it does not work. I am reverting back to the hardcoded value.

**Pull request merged**